File: src/nautilus-server/src/apps/polymarket/score.rs
// Polymarket prediction scoring math.
//
// Three job types, each producing an integer score in [0, 100] the competition contract records
// (record_score) and ranks:
//
//   resolution  -> 100 if the agent's YES/NO equals the market's resolved winner, else 0.
//   event       -> coverage-weighted: round(correct_guesses / total_markets_in_event * 100), so
//                  guessing MORE of the event's markets correctly scores higher and one lucky
//                  guess cannot reach 100.
//   price       -> Brier: round(100 * (1 - (guess - actual)^2)) over probabilities in [0, 1].
//
// All three are pure + deterministic. The price curve runs in basis-point integer space so the
// build stays reproducible (no float scoring), mirroring the finance scorers' integer discipline.

use std::collections::BTreeMap;

/// One guess in an event job: a market id and the agent's predicted outcome.
#[derive(Debug, Clone, serde::Deserialize, PartialEq, Eq)]
pub struct Guess {
    pub market_id: String,
    pub outcome: String,
}

/// Compare two outcome labels (e.g. "YES" vs "Yes") case-insensitively, ignoring surrounding space.
fn outcome_eq(a: &str, b: &str) -> bool {
    a.trim().eq_ignore_ascii_case(b.trim())
}

/// Resolution (Job #1): 100 when the agent's outcome matches the resolved winner, else 0.
pub fn score_resolution(agent_outcome: &str, winner: &str) -> u8 {
    if outcome_eq(agent_outcome, winner) {
        100
    } else {
        0
    }
}

/// Event (Job #2), coverage-weighted. `winners` maps each RESOLVED market id in the event to its
/// winning outcome. `total_markets` is the number of markets in the whole event (the denominator,
/// so partial coverage caps the score). A guess for a market not in `winners` scores nothing.
/// `total_markets == 0` is a degenerate event -> 0.
pub fn score_event(guesses: &[Guess], winners: &BTreeMap<String, String>, total_markets: usize) -> u8 {
    if total_markets == 0 {
        return 0;
    }
    let correct = guesses
        .iter()
        .filter(|g| matches!(winners.get(&g.market_id), Some(w) if outcome_eq(&g.outcome, w)))
        .count();
    // round(correct / total * 100), clamped to 100 (a duplicate guess can't exceed coverage).
    let scaled = (correct as u64 * 100 + total_markets as u64 / 2) / total_markets as u64;
    scaled.min(100) as u8
}

/// Convert a probability in [0, 1] to integer basis points in [0, 10000] (rounded, clamped).
fn to_bps(p: f64) -> i64 {
    if !p.is_finite() {
        return 0;
    }
    let bps = (p * 10_000.0).round() as i64;
    bps.clamp(0, 10_000)
}

/// Price (Job #3), Brier in basis-point space: 100 * (1 - d^2) with d = |guess - actual| in [0, 1].
/// d^2 = (d_bps / 10000)^2, so score = 100 - d_bps^2 / 1_000_000 (integer), clamped to [0, 100].
pub fn score_price(guess: f64, actual: f64) -> u8 {
    let d = (to_bps(guess) - to_bps(actual)).unsigned_abs();
    let penalty = (d * d) / 1_000_000; // in [0, 100]
    (100u64.saturating_sub(penalty)).min(100) as u8
}

#[cfg(test)]
mod test {
    use super::*;

    fn winners(pairs: &[(&str, &str)]) -> BTreeMap<String, String> {
        pairs.iter().map(|&(k, v)| (k.to_string(), v.to_string())).collect()
    }
    fn guesses(pairs: &[(&str, &str)]) -> Vec<Guess> {
        pairs
            .iter()
            .map(|&(m, o)| Guess { market_id: m.to_string(), outcome: o.to_string() })
            .collect()
    }

    #[test]
    fn resolution_is_all_or_nothing_case_insensitive() {
        assert_eq!(score_resolution("YES", "Yes"), 100);
        assert_eq!(score_resolution(" no ", "No"), 100);
        assert_eq!(score_resolution("YES", "No"), 0);
    }

    #[test]
    fn event_is_coverage_weighted() {
        let w = winners(&[("m1", "Yes"), ("m2", "No"), ("m3", "Yes"), ("m4", "No")]);
        // 3 correct out of 4 total markets -> 75.
        let g = guesses(&[("m1", "Yes"), ("m2", "No"), ("m3", "Yes"), ("m4", "Yes")]);
        assert_eq!(score_event(&g, &w, 4), 75);
        // One correct guess out of 4 total markets -> 25 (cannot reach 100 with a single guess).
        assert_eq!(score_event(&guesses(&[("m1", "Yes")]), &w, 4), 25);
        // All four correct -> 100.
        let all = guesses(&[("m1", "Yes"), ("m2", "No"), ("m3", "Yes"), ("m4", "No")]);
        assert_eq!(score_event(&all, &w, 4), 100);
    }

    #[test]
    fn event_ignores_guesses_for_unknown_markets() {
        let w = winners(&[("m1", "Yes"), ("m2", "No")]);
        // "m9" is not a resolved market in the event -> contributes nothing.
        let g = guesses(&[("m1", "Yes"), ("m9", "Yes")]);
        assert_eq!(score_event(&g, &w, 2), 50);
        assert_eq!(score_event(&[], &w, 2), 0);
        assert_eq!(score_event(&g, &w, 0), 0);
    }

    #[test]
    fn price_brier_rewards_closeness() {
        assert_eq!(score_price(0.62, 0.62), 100); // exact
        assert_eq!(score_price(0.5, 0.5), 100);
        // off by 0.10 -> 100 - 1000^2/1e6 = 100 - 1 = 99.
        assert_eq!(score_price(0.60, 0.50), 99);
        // off by 0.50 -> 100 - 5000^2/1e6 = 100 - 25 = 75.
        assert_eq!(score_price(1.0, 0.5), 75);
        // off by 1.0 -> 100 - 10000^2/1e6 = 0.
        assert_eq!(score_price(1.0, 0.0), 0);
    }

    #[test]
    fn price_clamps_out_of_range_inputs() {
        // Inputs outside [0,1] are clamped before scoring.
        assert_eq!(score_price(1.5, 1.0), 100);
        assert_eq!(score_price(-0.2, 0.0), 100);
        assert_eq!(score_price(f64::NAN, 0.0), 100); // NaN -> 0 bps == actual 0
    }
}

File: src/evaluators/polymarket/score.ts
/**
 * polymarket scoring — port of the Sui engine's `apps/polymarket/score.rs`.
 *
 * Three evaluators, each producing an integer score in [0,100]:
 *
 *   resolution — 100 if the agent's outcome equals the market's resolved winner, else 0.
 *   event      — coverage-weighted: correct guesses over the number of markets in the WHOLE
 *                event, so one lucky guess cannot reach 100 and answering more of the event pays.
 *   price      — Brier over probabilities, in basis-point integer space.
 *
 * Pure and integer-only, matching the price scorers' discipline. The float inputs are converted to
 * basis points immediately and never touched again.
 *
 * THESE SCORES CANNOT CURRENTLY BE SETTLED. Every settlement path in `JobEscrow` and
 * `SealedCompetition` calls `FtsoLib.checkGroundTruth`, which requires an FTSO anchor-feed Merkle
 * proof whose normalized value equals the signed `groundTruthValue`. A Polymarket outcome has no
 * FTSO feed, so there is no proof to attach. Attaching an unrelated feed's proof would satisfy the
 * contract and be a lie, so the engine refuses instead — see `evaluators/index.ts`. Landing these
 * for real needs an FDC `Web2Json` verification path in `contracts`, which is a different repo and
 * a redeploy.
 */

/** Outcome labels are compared case-insensitively, ignoring surrounding space ("YES" == "Yes"). */
export function outcomeEq(a: string, b: string): boolean {
    return a.trim().toLowerCase() === b.trim().toLowerCase();
}

export interface Guess {
    readonly marketId: string;
    readonly outcome: string;
}

/** 100 when the agent's outcome matches the resolved winner, else 0. */
export function scoreResolution(agentOutcome: string, winner: string): number {
    return outcomeEq(agentOutcome, winner) ? 100 : 0;
}

/**
 * Coverage-weighted event score.
 *
 * `winners` maps each RESOLVED market id to its winning outcome. `totalMarkets` is the count for
 * the whole event, and it is the denominator on purpose: partial coverage caps the score, so an
 * agent that answers one market of ten correctly scores 10, not 100.
 */
export function scoreEvent(
    guesses: readonly Guess[],
    winners: ReadonlyMap<string, string>,
    totalMarkets: number,
): number {
    if (totalMarkets <= 0) return 0;
    let correct = 0;
    for (const guess of guesses) {
        const winner = winners.get(guess.marketId);
        if (winner !== undefined && outcomeEq(guess.outcome, winner)) correct += 1;
    }
    // Round half up, exactly as the Rust `(correct * 100 + total / 2) / total` did.
    const scaled = Math.floor((correct * 100 + Math.floor(totalMarkets / 2)) / totalMarkets);
    return Math.min(100, scaled);
}

/** A probability in [0,1] as integer basis points. Non-finite becomes 0; out of range clamps. */
export function toBps(probability: number): number {
    if (!Number.isFinite(probability)) return 0;
    return Math.min(10_000, Math.max(0, Math.round(probability * 10_000)));
}

/**
 * Brier score over probabilities: `100 - d^2 / 1e6` with `d` the basis-point error.
 *
 * Much flatter than the up/down Brier by design: an error of 0.10 costs one point, 0.50 costs 25,
 * and only a completely inverted call reaches 0.
 */
export function scorePrice(guess: number, actual: number): number {
    const d = Math.abs(toBps(guess) - toBps(actual));
    const penalty = Math.floor((d * d) / 1_000_000);
    return Math.min(100, Math.max(0, 100 - penalty));
}

/** `guesses` arrives as JSON encoded into a string field, the Sui convention. */
export function parseGuesses(raw: string | undefined): Guess[] | undefined {
    if (raw === undefined) return undefined;
    let parsed: unknown;
    try {
        parsed = JSON.parse(raw);
    } catch {
        return undefined;
    }
    if (!Array.isArray(parsed)) return undefined;
    const out: Guess[] = [];
    for (const entry of parsed) {
        if (!entry || typeof entry !== 'object') return undefined;
        const row = entry as Record<string, unknown>;
        // Accept both spellings: the Sui payloads used snake_case.
        const marketId = row['marketId'] ?? row['market_id'];
        const outcome = row['outcome'];
        if (typeof marketId !== 'string' || typeof outcome !== 'string') return undefined;
        out.push({ marketId, outcome });
    }
    return out;
}
